File: frontend/app/lib/study.ts
/**
 * Browser-side study + continue-reading client (RFC-004 Cycle E, HOME-01/09/11).
 *
 * Thin helpers the Home surface uses to read the adherence rollup and the
 * continue-reading hero from FastAPI *through the same-origin Next.js proxy*
 * (`/api/...`, ADR-017) — never cross-origin. The HttpOnly session cookie rides
 * along automatically (`credentials: "same-origin"`), so this code never reads or
 * holds the session token. Both are read-only GETs, so they carry no CSRF token.
 *
 * The day boundary is user-local: `getStudyDays` echoes the caller's IANA zone in
 * `X-Client-Timezone` so the window ends at their local today (HOME-09/11). The
 * zone comes from `clientTimezone()`, which the reading-position and
 * review-submission writers reuse to date their rollups the same way. When the
 * zone cannot be resolved the header is simply omitted — never sent as the string
 * `"undefined"` — and the backend falls back to UTC (AD-152, HOME-09).
 *
 * FastAPI stays authoritative for auth, user-scoping, and the read-time
 * `studied_last_14`. The response types mirror the backend views in `web/study.py`
 * exactly so the two never drift.
 */

/** The tz header the day boundary rides on (HOME-09). */
const CLIENT_TIMEZONE_HEADER = "X-Client-Timezone";

/** One user-local day of activity, mirroring the backend `StudyDayView`. */
export type StudyDayView = {
  day: string;
  reviews_count: number;
  reading_updates: number;
};

/** The adherence read model, mirroring the backend `StudySummaryView`. */
export type StudySummaryView = {
  days: StudyDayView[];
  studied_last_14: number;
};

/** The continue-reading hero, mirroring the backend `ContinueReadingView`. */
export type ContinueReadingView = {
  source_id: string;
  source_title: string;
  chapter_title: string;
  percent: number;
  updated_at: string;
};

/**
 * The caller's IANA time zone from the browser, or `undefined` when it cannot be
 * resolved (AD-152). Callers attach it as `X-Client-Timezone` only when defined,
 * so an unavailable `Intl` degrades to the server's UTC fallback rather than
 * sending a bogus zone. Never throws.
 */
export function clientTimezone(): string | undefined {
  try {
    return Intl.DateTimeFormat().resolvedOptions().timeZone || undefined;
  } catch {
    return undefined;
  }
}

/**
 * Fetch the caller's most-recent reading position as the Home hero, or `null`
 * when they have no positions yet (200 empty shape → the hero's pick-a-book
 * state, HOME-02). Read-only GET, so no CSRF token. Throws a readable error on a
 * non-OK response (e.g. 401 unauthenticated) so the hero can show its own error.
 */
export async function getContinueReading(
  fetchImpl: typeof fetch = fetch,
): Promise<ContinueReadingView | null> {
  const res = await fetchImpl("/api/reading/continue", {
    method: "GET",
    credentials: "same-origin",
  });
  if (!res.ok) {
    throw await toStudyError(res, "Could not load your reading progress.");
  }
  return (await res.json()) as ContinueReadingView | null;
}

/**
 * Fetch the caller's study-day window plus `studied_last_14` (HOME-11). The
 * optional `window` (days) is passed through to the endpoint; when omitted the
 * server applies its 84-day default. Echoes `X-Client-Timezone` so the window
 * ends at the caller's local today, omitting it when the zone is unavailable
 * (HOME-09). Read-only GET, so no CSRF token. Throws a readable error on a
 * non-OK response.
 */
export async function getStudyDays(
  window?: number,
  fetchImpl: typeof fetch = fetch,
): Promise<StudySummaryView> {
  const query = window !== undefined ? `?window=${window}` : "";
  const headers: Record<string, string> = {};
  const tz = clientTimezone();
  if (tz) {
    headers[CLIENT_TIMEZONE_HEADER] = tz;
  }
  const res = await fetchImpl(`/api/study/days${query}`, {
    method: "GET",
    credentials: "same-origin",
    headers,
  });
  if (!res.ok) {
    throw await toStudyError(res, "Could not load your study activity.");
  }
  return (await res.json()) as StudySummaryView;
}

/**
 * Build an Error from a non-OK response, preferring the backend's detail. A
 * non-string detail (e.g. a 422 validation list) falls back to the readable
 * message, mirroring `quiz.ts`.
 */
async function toStudyError(res: Response, fallback: string): Promise<Error> {
  try {
    const body = (await res.json()) as { detail?: unknown };
    return new Error(typeof body.detail === "string" ? body.detail : fallback);
  } catch {
    return new Error(fallback);
  }
}

File: backend/app/infrastructure/web/study.py
"""Study + continue-reading router — the read side of the Home surface (RFC-004 Cycle E).

Thin FastAPI adapters over the framework-free ``GetStudySummary`` and ``ContinueReading``
services. Both are plain authenticated GETs (no CSRF, no job). The optional
``X-Client-Timezone`` header sets the study-day boundary for the summary; the continue
endpoint has no day logic and does not read it.

Contract:
- ``GET /api/study/days?window=`` → 200 window rows + ``studied_last_14``; ``window``
  defaults to 84, bounds 7..365 (else 422). No session → 401 (HOME-11).
- ``GET /api/reading/continue`` → 200 hero shape, or ``null`` when the caller has no
  reading positions. No session → 401 (HOME-01/02).
"""

from __future__ import annotations

from datetime import date, datetime
from typing import Annotated
from uuid import UUID

from fastapi import APIRouter, Depends, Header, Query
from pydantic import BaseModel

from app.application.study import ContinueReading, GetStudySummary
from app.domain.entities import ContinuePoint, StudySummary, User
from app.infrastructure.web.dependencies import (
    get_authenticated_user,
    get_continue_reading,
    get_study_summary,
)

router = APIRouter(tags=["study"])


class StudyDayView(BaseModel):
    """One user-local day of activity with its per-kind counters (HOME-11)."""

    day: date
    reviews_count: int
    reading_updates: int


class StudySummaryView(BaseModel):
    """The adherence read model: window rows + the 14-day adherence count (HOME-11/12)."""

    days: list[StudyDayView]
    studied_last_14: int

    @classmethod
    def from_summary(cls, summary: StudySummary) -> StudySummaryView:
        return cls(
            days=[
                StudyDayView(
                    day=d.day,
                    reviews_count=d.reviews_count,
                    reading_updates=d.reading_updates,
                )
                for d in summary.days
            ],
            studied_last_14=summary.studied_last_14,
        )


class ContinueReadingView(BaseModel):
    """The continue-reading hero: where to resume in one click (HOME-01/03)."""

    source_id: UUID
    source_title: str
    chapter_title: str
    percent: float
    updated_at: datetime

    @classmethod
    def from_point(cls, point: ContinuePoint) -> ContinueReadingView:
        return cls(
            source_id=point.source_id,
            source_title=point.source_title,
            chapter_title=point.chapter_title,
            percent=float(point.percent),
            updated_at=point.updated_at,
        )


@router.get("/api/study/days")
def get_study_days(
    user: Annotated[User, Depends(get_authenticated_user)],
    service: Annotated[GetStudySummary, Depends(get_study_summary)],
    window: Annotated[int, Query(ge=7, le=365)] = 84,
    client_tz: Annotated[str | None, Header(alias="X-Client-Timezone")] = None,
) -> StudySummaryView:
    """Return the caller's study-day window + ``studied_last_14`` (200; 401; 422).

    ``window`` defaults to 84 (12 weeks) and is bounded 7..365 by Pydantic — out of
    range → 422. The day boundary follows the optional ``X-Client-Timezone`` header
    (silent UTC fallback). Everything is derived at read time; nothing is stored.
    """
    summary = service(user=user, window=window, tz=client_tz)
    return StudySummaryView.from_summary(summary)


@router.get("/api/reading/continue")
def get_continue_reading_endpoint(
    user: Annotated[User, Depends(get_authenticated_user)],
    service: Annotated[ContinueReading, Depends(get_continue_reading)],
) -> ContinueReadingView | None:
    """Return the caller's most-recent reading position as the hero, or ``null`` (200; 401).

    ``ContinueReading`` reads the caller's most-recent position (user-scoped in SQL) and
    resolves its chapter title; a caller with no positions gets ``null`` (200) so the hero
    renders its pick-a-book empty state.
    """
    point = service(user=user)
    return ContinueReadingView.from_point(point) if point is not None else None
